File: olorin-media/bayit-plus/web/src/services/featureFlags.ts
/**
 * Feature Flags Service
 * Centralized feature flag management for frontend
 */

import { api } from './api';

/**
 * Feature flags interface
 */
export interface FeatureFlags {
  new_player: boolean;
  dark_mode: boolean;
  offline_mode: boolean;
  recommendations: boolean;
  social_features: boolean;
  live_chat: boolean;
  analytics_v2: boolean;
  scene_search: boolean;
  [key: string]: boolean;
}

/**
 * In-memory cache for feature flags
 */
let featureFlagsCache: FeatureFlags | null = null;
let cacheTimestamp: number = 0;
const CACHE_TTL_MS = 5 * 60 * 1000; // 5 minutes

/**
 * Default feature flags (fallback if API unavailable)
 */
const DEFAULT_FEATURE_FLAGS: FeatureFlags = {
  new_player: true,
  dark_mode: true,
  offline_mode: false,
  recommendations: true,
  social_features: false,
  live_chat: true,
  analytics_v2: false,
  scene_search: true,
};

/**
 * Get current feature flags from backend
 */
export async function getFeatureFlags(): Promise<FeatureFlags> {
  // Check cache first
  const now = Date.now();
  if (featureFlagsCache && now - cacheTimestamp < CACHE_TTL_MS) {
    return featureFlagsCache;
  }

  try {
    // Public endpoint (no auth required)
    const response = await api.get<FeatureFlags>('/api/v1/admin/settings/feature-flags/public');
    featureFlagsCache = response.data;
    cacheTimestamp = now;
    return response.data;
  } catch (error) {
    console.warn('Failed to fetch feature flags, using cache or defaults:', error);

    // Return cache if available
    if (featureFlagsCache) {
      return featureFlagsCache;
    }

    // Last resort: default flags
    return DEFAULT_FEATURE_FLAGS;
  }
}

/**
 * Check if a specific feature is enabled
 */
export async function isFeatureEnabled(featureName: string): Promise<boolean> {
  const flags = await getFeatureFlags();
  return flags[featureName] ?? false;
}

/**
 * Check multiple features
 */
export async function checkMultipleFeatures(
  featureNames: string[],
  requireAll: boolean = true
): Promise<boolean> {
  const flags = await getFeatureFlags();
  const statuses = featureNames.map((name) => flags[name] ?? false);

  if (requireAll) {
    return statuses.every((status) => status === true);
  } else {
    return statuses.some((status) => status === true);
  }
}

/**
 * Get list of all enabled features
 */
export async function getEnabledFeatures(): Promise<string[]> {
  const flags = await getFeatureFlags();
  return Object.entries(flags)
    .filter(([_, enabled]) => enabled)
    .map(([name]) => name);
}

/**
 * Clear the feature flags cache (force refresh on next call)
 */
export function clearCache(): void {
  featureFlagsCache = null;
  cacheTimestamp = 0;
}

/**
 * Subscribe to feature flag changes (polling)
 */
export function subscribeToFeatureFlags(
  callback: (flags: FeatureFlags) => void,
  intervalMs: number = 60000 // 1 minute
): () => void {
  let intervalId: NodeJS.Timeout | null = null;

  const poll = async () => {
    clearCache(); // Force refresh
    const flags = await getFeatureFlags();
    callback(flags);
  };

  // Initial fetch
  poll();

  // Set up polling
  intervalId = setInterval(poll, intervalMs);

  // Return cleanup function
  return () => {
    if (intervalId) {
      clearInterval(intervalId);
      intervalId = null;
    }
  };
}

/**
 * Feature flag utility for conditional rendering
 */
export const featureFlags = {
  get: getFeatureFlags,
  isEnabled: isFeatureEnabled,
  checkMultiple: checkMultipleFeatures,
  getEnabled: getEnabledFeatures,
  clearCache,
  subscribe: subscribeToFeatureFlags,
};

File: olorin-media/bayit-plus/backend/app/services/feature_flags.py
"""
Feature Flag Service
Centralized feature flag management and checking
"""

import logging
import os
from typing import Dict, Optional

from fastapi import HTTPException, status

from app.models.admin import SystemSettings

logger = logging.getLogger(__name__)

# In-memory cache for feature flags (refreshed periodically)
_feature_flags_cache: Optional[Dict[str, bool]] = None


async def get_feature_flags() -> Dict[str, bool]:
    """
    Get current feature flags from database.

    Returns:
        Dictionary of feature flag names to their enabled status
    """
    global _feature_flags_cache

    # Try database first
    settings = await SystemSettings.find_one(SystemSettings.key == "system_settings")

    if settings and settings.feature_flags:
        _feature_flags_cache = settings.feature_flags
        return settings.feature_flags

    # Fallback to cache if DB unavailable
    if _feature_flags_cache:
        logger.warning("Using cached feature flags (database unavailable)")
        return _feature_flags_cache

    # Last resort: default flags
    logger.warning("Using default feature flags (no database or cache)")
    return {
        "new_player": True,
        "dark_mode": True,
        "offline_mode": False,
        "recommendations": True,
        "social_features": False,
        "live_chat": True,
        "analytics_v2": False,
        "scene_search": True,
    }


async def is_feature_enabled(feature_name: str) -> bool:
    """
    Check if a feature flag is enabled.

    Checks in this order:
    1. Environment variable FEATURE_{NAME}_ENABLED (override)
    2. Database SystemSettings.feature_flags
    3. Cached flags
    4. Default to False for unknown features

    Args:
        feature_name: Name of the feature (e.g., "scene_search")

    Returns:
        True if enabled, False otherwise
    """
    # Check environment variable override (deployment-level control)
    env_var = f"FEATURE_{feature_name.upper()}_ENABLED"
    env_value = os.getenv(env_var)
    if env_value is not None:
        enabled = env_value.lower() in ("true", "1", "yes")
        logger.info(
            f"Feature '{feature_name}' controlled by env var {env_var}: {enabled}"
        )
        return enabled

    # Check database/cache
    flags = await get_feature_flags()
    enabled = flags.get(feature_name, False)

    logger.debug(f"Feature '{feature_name}' enabled: {enabled}")
    return enabled


async def require_feature(feature_name: str) -> None:
    """
    Verify a feature is enabled. Raises HTTPException if not.

    Use as FastAPI dependency:
        @router.get("/endpoint")
        async def my_endpoint(
            _: None = Depends(lambda: require_feature("scene_search"))
        ):
            ...

    Args:
        feature_name: Feature to check

    Raises:
        HTTPException: 503 Service Unavailable if feature disabled
    """
    if not await is_feature_enabled(feature_name):
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail={
                "error": "feature_disabled",
                "message": f"The '{feature_name}' feature is currently unavailable.",
                "feature": feature_name,
            },
        )


async def check_multiple_features(
    *feature_names: str, require_all: bool = True
) -> bool:
    """
    Check multiple feature flags.

    Args:
        *feature_names: Feature names to check
        require_all: If True, all must be enabled. If False, any can be enabled.

    Returns:
        True if check passes, False otherwise
    """
    flags = await get_feature_flags()
    statuses = [flags.get(name, False) for name in feature_names]

    if require_all:
        return all(statuses)
    else:
        return any(statuses)


async def get_enabled_features() -> list[str]:
    """
    Get list of all currently enabled features.

    Returns:
        List of enabled feature names
    """
    flags = await get_feature_flags()
    return [name for name, enabled in flags.items() if enabled]


def clear_cache() -> None:
    """Clear the in-memory feature flags cache."""
    global _feature_flags_cache
    _feature_flags_cache = None
    logger.info("Feature flags cache cleared")
